src/hooks/sync-sites-context.tsx

```diff
@@ -0,0 +1,212 @@
+import { useI18n } from '@wordpress/react-i18n';
+import React, { createContext, useContext, useState, useCallback, useEffect, useMemo } from 'react';
+import { getIpcApi } from '../lib/get-ipc-api';
+import { useAuth } from './use-auth';
+import { useImportExport } from './use-import-export';
+import { PullStateProgressInfo, useSyncStatesProgressInfo } from './use-sync-states-progress-info';
+
+export type SiteBackupState = {
+	backupId: string | null;
+	status: PullStateProgressInfo;
+	downloadUrl: string | null;
+	selectedSite: SiteDetails;
+};
+
+type SyncSitesContextType = {
+	pullStates: Record< number, SiteBackupState >;
+	pullSite: ( remoteSiteId: number, selectedSite: SiteDetails ) => Promise< void >;
+	isAnySitePulling: boolean;
+};
+
+const SyncSitesContext = createContext< SyncSitesContextType | undefined >( undefined );
+
+export function SyncSitesProvider( { children }: { children: React.ReactNode } ) {
+	const { pullStates, pullSite, isAnySitePulling } = useSyncPull();
+
+	return (
+		<SyncSitesContext.Provider value={ { pullStates, pullSite, isAnySitePulling } }>
+			{ children }
+		</SyncSitesContext.Provider>
+	);
+}
+
+function useSyncPull() {
+	const { __ } = useI18n();
+	const { client } = useAuth();
+	const [ pullStates, setPullStates ] = useState< Record< number, SiteBackupState > >( {} );
+	const { importFile } = useImportExport();
+	const { pullStatesProgressInfo, isKeyPulling } = useSyncStatesProgressInfo();
+
+	const pullSite = useCallback(
+		async ( remoteSiteId: number, selectedSite: SiteDetails ) => {
+			if ( ! client ) {
+				return;
+			}
+			setPullStates( ( prevStates ) => ( {
+				...prevStates,
+				[ remoteSiteId ]: {
+					backupId: null,
+					status: pullStatesProgressInfo[ 'in-progress' ],
+					downloadUrl: null,
+					selectedSite,
+				},
+			} ) );
+			try {
+				const response = await client.req.post< { success: boolean; backup_id: string } >( {
+					path: `/sites/${ remoteSiteId }/studio-app/sync/backup`,
+					apiNamespace: 'wpcom/v2',
+				} );
+
+				if ( response.success ) {
+					setPullStates( ( prevStates ) => ( {
+						...prevStates,
+						[ remoteSiteId ]: { ...prevStates[ remoteSiteId ], backupId: response.backup_id },
+					} ) );
+				} else {
+					throw new Error( 'Pull request failed' );
+				}
+			} catch ( error ) {
+				console.error( error );
+				setPullStates( ( prevStates ) => ( {
+					...prevStates,
+					[ remoteSiteId ]: {
+						...prevStates[ remoteSiteId ],
+						status: pullStatesProgressInfo.failed,
+					},
+				} ) );
+				getIpcApi().showErrorMessageBox( {
+					title: __( 'Failed to pull backup' ),
+					message: error instanceof Error ? error.message : __( 'Unknown error' ),
+				} );
+			}
+		},
+		[ __, client, pullStatesProgressInfo ]
+	);
+
+	const onBackupCompleted = useCallback(
+		async ( remoteSiteId: number, downloadUrl: string, selectedSite: SiteDetails ) => {
+			setPullStates( ( prevStates ) => ( {
```

Comment:
I see that we are repeating a similar logic (besides changing the status) for a couple of times. What do you think about extracting `setPullStates` into some sort of a helper function?

Reply:
> I see your eye is sharper than ever  

😂